### PR TITLE
Add metadata headers when uploading blobs

### PR DESCRIPTION
**Pull request by noamt**
Adding the following metadata headers to the upload request:
* Object ID
* Original file name
* Git branch
* Git ref

Adding these headers will benefit smart remote repositories and will help improve BLOB management and tracking.

### CODE REVIEW

**Comment by technoweenie** [comment]
Nice idea, but I prefer adding it to the batch api json document as suggested in #845. Why would you want this on the storage upload requests, and not the batch API? LFS was designed to use a custom LFS API against a potentially generic storage backend (such as S3), which would ignore the extra headers. Would implementing #845 be enough, or do you _need_ these headers?

Calling `git.CurrentRef()` AND `git.CurrentBranch()` is unnecessary. `CurrentRef()` returns more information. Both functions also invoke `git rev-parse`. We don't want to add 1 or 2 extra redundant `git rev-parse` calls to every file that gets uploaded. Might be better to add extra properties to `*Uploadable` instead, that get passed in somewhere at a higher level.

Super nitpicky: the `X-` prefix on custom headers is unnecessary. I'd probably go with header keys like `Git-Lfs-*`.

File: lfs/client.go [hunk 1]
```diff
@@ -795,3 +820,12 @@ func setErrorHeaderContext(err error, prefix string, head http.Header) {
...
 }
+
+func PrintError(err error, format string, args ...interface{}) {
+	line := format
+	if len(args) > 0 {
+		line = fmt.Sprintf(format, args...)
+	}
+	fmt.Fprintln(ErrorWriter, line)
+	fmt.Fprintln(ErrorWriter, err.Error())
+}
```
Comment (technoweenie) [inline lfs/client.go]: This is the type of error collecting thing that the `commands` package should take care of. The `lfs` package is more of a library with functions that should bubble errors up. The `commands` package is the CLI layer that prints and logs errors.

Comment (noamt) [inline lfs/client.go]: I understand. My reason for doing this was thinking that the headers are not critical to the operation, so the operation shouldn't fail if any problems come up.

**Comment by noamt** [comment]
Thanks for the feedback, you're right. It's more fitting in the batch API.
This also means that the batch API responses can be smarter in handling blob/path duplication.

I'd still like to have this feature, so if I understand you correctly you'd like to:
* Move this info to the batch API JSON.
* Drop the invocation of ```git.CurrentBranch()```.
* Move the collection of git information up the stack.

Should I close this and open a new pull request?

File: lfs/client.go (hunk 1 above)
Comment (technoweenie) [inline lfs/client.go]: We have a `tracerx` package that logs stuff if you have pass `GIT_TRACE=1` to git lfs. This actually piggy backs on a core git feature, so running with git tracing on will show you all kinds of stuff :)

**Comment by technoweenie** [comment]
> *  Move this info to the batch API JSON.
> * Drop the invocation of git.CurrentBranch().
> * Move the collection of git information up the stack.

:+1: If you could update the batch api docs to go with it, that'd be cool too. Depends how involved you want to get with the project. Also, as you're messing around in the lfs client code, there IS an issue focused on cleaning it up: #779. So, don't hesitate to call out anything gross or weird.

> Should I close this and open a new pull request?

You're welcome to do whatever you're most comfortable with. If it was me, I'd just keep hacking on this PR. 

I appreciate you taking the time to dig in! Feel free to jump in [chat](https://gitter.im/github/git-lfs) if you want to talk. It's a very low volume channel, especially during a US holiday week like this week :)

**Comment by noamt** [comment]
Cheers, I'll get right on it.

**Comment by noamt** [comment]
OK, so I've shuffled things around and now each object of the batch request will contain the original file name and the batch request itself will contain the git revision.

So now I'd like to resolve and pass down the current branch. Assuming I'm at ```commands/command_pre_push.go:72```, what do you think would be the best way to do that?

File: lfs/client.go [hunk 2]
```diff
@@ -204,12 +207,12 @@ func (b *byteCloser) Close() error {
 	return nil
 }
 
-func Batch(objects []*ObjectResource, operation string) ([]*ObjectResource, error) {
```
Comment (technoweenie) [inline lfs/client.go]: I wonder if we should introduce some kind struct here instead?

```go
type BatchRequest struct {
  Operation string
  Rev string
}

// example pseudo code
ref, _ := git.CurrentRef()
req := lfs.NewRequest("download", objects)
req.Ref = ref.Name
lfs.Batch(req)
```

**Comment by technoweenie** [comment]
> So now I'd like to resolve and pass down the current branch. Assuming I'm at `commands/command_pre_push.go:72`, what do you think would be the best way to do that?

Probably call `git.CurrentRef()` in the command, and pass it to the `uploadQueue`. Maybe some kind of generic struct to represent the current state of the git repo, like the current ref. Or maybe all it needs is the ref?

Getting the filename may be tricky though. It'd have to come from `lfs.ScanRefs()`, and then passed to the uploadable.

```go
// example
u, _ := lfs.NewUploadable(pointer.Oid, pointer.Name)
// it is possible for a single pointer to match multiple filenames
u.Filenames = pointer.Filenames
```

**Comment by noamt** [comment]
So yeah I've actually got the file name from the pointer and I enter it into each individual object so that's fine.

**Comment by noamt** [comment]
OK I think I nailed it. Added a new struct object ```TransferMetadata``` to the queue instead of polluting all these methods as the amount of data will also probably grow.
If it looks good I'll continue on to add tests and documentation.

**Comment by noamt** [comment]
Done and done

**Comment by noamt** [comment]
Though it's possible I don't think I've come across many instances of that case (local branch name != remote branch name) out in the wild , but I agree that it's better

**Comment by dluksza** [comment]
'ref' property can be used in Gerrit for branch based permissions, it would create a security hole if it is name of local branch not destination branch :)

**Comment by noamt** [comment]
Fixed. Thanks for the feedback @dluksza

**Pull request closed without merging**